server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRecipientRewriteTableMailetTest.java

```diff
@@ -0,0 +1,347 @@
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader(RFC2822Headers.RETURN_PATH)).containsOnly("<>");
+    }
+
+    @Test
+    public void serviceShouldResetValueOfMessageHeaderWhenNotNullSender() throws Exception {
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .sender(MailAddressFixture.ANY_AT_JAMES)
+            .mimeMessage(message)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES))
+            .build();
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader(RFC2822Headers.RETURN_PATH)).containsOnly("<" + MailAddressFixture.ANY_AT_JAMES + ">");
+    }
+
+    @Test
+    public void serviceShouldWriteLogAndSendMailToAllErrorRecipients() throws Exception {
+        mailetConfig = new FakeMailetConfig("Test", mockMailetContext);
+
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES))
+            .build();
+
+        mailet.service(mail);
+
+        verify(mockMailetContext, times(1)).log(eq(LogLevel.INFO), eq("Error while process mail."), any(MessagingException.class));
+        verify(mockMailetContext).sendMail(eq(mail.getSender()), eq(ImmutableList.of(MailAddressFixture.OTHER_AT_JAMES)), eq(message), eq(Mail.ERROR));
+        verifyNoMoreInteractions(mockMailetContext);
+    }
+
+    @Test
+    public void serviceShouldNotSendMailWhenNoErrorRecipients() throws Exception {
+        mailetConfig = new FakeMailetConfig("Test", mockMailetContext);
+
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(ImmutableList.of(MailAddressFixture.ANY_AT_JAMES, nonDomainWithDefaultLocal))
+            .build();
+
+        mailet.service(mail);
+
+        verifyNoMoreInteractions(mockMailetContext);
+    }
+
+    @Test
+    public void serviceShouldResetMailStateToGhostWhenCanNotBuildNewRecipient() throws Exception {
+        mailet.init(mailetConfig);
+        mail = FakeMail.builder()
+            .mimeMessage(message)
+            .recipients(MailAddressFixture.OTHER_AT_JAMES, nonDomainWithDefaultLocal)
+            .build();
+
+        mailet.service(mail);
+
+        assertThat(mail.getState()).isEqualTo(Mail.GHOST);
+        assertThat(mail.getRecipients()).isEmpty();
+    }
+
+    @Test(expected = MessagingException.class)
+    public void handleMappingsShouldThrowExceptionWhenMappingsContainAtLeastOneNoneDomainObjectButCannotGetDefaultDomain() throws Exception {
+        when(domainList.getDefaultDomain()).thenThrow(DomainListException.class);
+        mailet.setDomainList(domainList);
+        mappings = MappingsImpl.fromRawString(MailAddressFixture.ANY_AT_JAMES.toString() + "," + NONEDOMAIN + "," + MailAddressFixture.OTHER_AT_JAMES.toString());
```

Comment:
MappingsImpl.fromRawString was written by @aduprat to support old code without refactoring too much.

Unfortunately, this raw string is hard, even impossible to understand at first glance. Hopefully a builder would allow us to put a name on the different parts of the mapping and better understand your intent, at the negligeable cost of ~4 more lines of code. I agree with @mbaechler : please use builder here ;-)